src/main/java/com/emc/object/s3/S3Signer.java

```diff
@@ -0,0 +1,219 @@
+package com.emc.object.s3;
+
+import com.emc.object.s3.jersey.BucketFilter;
+import com.emc.object.s3.jersey.NamespaceFilter;
+import com.emc.object.s3.request.PresignedUrlRequest;
+import com.emc.object.util.RestUtil;
+import com.sun.jersey.api.client.ClientRequest;
+import org.apache.commons.codec.binary.Hex;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import javax.crypto.Mac;
+import javax.crypto.spec.SecretKeySpec;
+import java.net.MalformedURLException;
+import java.net.URI;
+import java.net.URL;
+import java.nio.charset.StandardCharsets;
+import java.security.InvalidKeyException;
+import java.security.MessageDigest;
+import java.security.NoSuchAlgorithmException;
+import java.util.*;
+
+public abstract class S3Signer {
+    protected static final Logger log = LoggerFactory.getLogger(S3Signer.class);
+
+    protected S3Config s3Config;
+    protected SortedSet<String> signedParameters;
+
+    S3Signer(S3Config s3Config) {
+        this.s3Config = s3Config;
+        signedParameters = new TreeSet<String>(Arrays.asList(
+                "acl", "torrent", "logging", "location", "policy", "requestPayment", "versioning",
+                "versions", "versionId", "notification", "uploadId", "uploads", "partNumber", "website",
+                "delete", "lifecycle", "tagging", "cors", "restore",
+                S3Constants.PARAM_RESPONSE_HEADER_CACHE_CONTROL,
+                S3Constants.PARAM_RESPONSE_HEADER_CONTENT_DISPOSITION,
+                S3Constants.PARAM_RESPONSE_HEADER_CONTENT_ENCODING,
+                S3Constants.PARAM_RESPONSE_HEADER_CONTENT_LANGUAGE,
+                S3Constants.PARAM_RESPONSE_HEADER_CONTENT_TYPE,
+                S3Constants.PARAM_RESPONSE_HEADER_EXPIRES,
+                S3Constants.PARAM_ENDPOINT,
+                S3Constants.PARAM_IS_STALE_ALLOWED));
+        if (s3Config.isSignMetadataSearch()) {
+            signedParameters.add(S3Constants.PARAM_QUERY);
+            signedParameters.add(S3Constants.PARAM_SEARCH_METADATA);
+        }
+    }
+
+    public abstract void sign(ClientRequest request, String resource, Map<String, String> parameters,
+                              Map<String, List<Object>> headers);
+
+    protected abstract String getSignature(String stringToSign, byte[] signingKey);
```

Comment:
need docs on these abstract methods - what is the method contract?  i.e. what happens if you pass in `null` to `signingKey`?

Reply:
Added java docs for these methods.